### PR TITLE
server: relax server connection dropping

### LINKED ISSUE #1337: lnd refuses inbound connections
lnd thinks that a node is connected even though the tcp socket is down. gossip-ping timeout doesn't help either.
also a mutex is hung somewhere because `updatechanpolicy` doesn't work, but most other commands work.
it doesn't seem to forward/send/receive any payments though.

```
2018-06-06 15:31:09.794 [WRN] SRVR: Received inbound connection from peer 03685a27a2d29f3f14d02dde544537317b624e39e540b6389fc6404dc3d452cb1d, but already connected, dropping conn
2018-06-06 15:36:10.048 [WRN] SRVR: Received inbound connection from peer 03685a27a2d29f3f14d02dde544537317b624e39e540b6389fc6404dc3d452cb1d, but already connected, dropping conn
2018-06-06 15:41:10.379 [WRN] SRVR: Received inbound connection from peer 03685a27a2d29f3f14d02dde544537317b624e39e540b6389fc6404dc3d452cb1d, but already connected, dropping conn
2018-06-06 15:46:10.664 [WRN] SRVR: Received inbound connection from peer 03685a27a2d29f3f14d02dde544537317b624e39e540b6389fc6404dc3d452cb1d, but already connected, dropping conn
```
```
$ sudo netstat -nap | grep 5.39.82.105
$ lncli connect 03685a27a2d29f3f14d02dde544537317b624e39e540b6389fc6404dc3d452cb1d@5.39.82.105:9735
[lncli] rpc error: code = Unknown desc = already connected to peer: 5.39.82.105:58642
$ sudo netstat -nap | grep 5.39.82.105
```

### PR DESCRIPTION
**Pull request by cfromknecht**
Think this will fix #1337 

Problem
======

We've recently seen some issues involving tight connections loops, which stem from different implementations having different policies for how to handle the case where both nodes successfully establish outgoing connections. The problem comes down to: what procedure should we use to choose between the incoming and outgoing connection.

The connection loops occur when both peers decide to terminate different connections, resulting in no usable transport, and both attempting to reconnect again. Without backoff strategies, it is possible this cycle can continue for some time if the network is somewhat reliable. The two strategies currently deployed AFIAA are:
1. Reject incoming if already have outgoing.
2. Reject if remote pubkey is larger.

Let's say we have nodes with pubkeys A and B, such that pubkey A < B. The possible strategy combinations are:
1. Both use 1.
2. A uses 1, B uses 2, WLOG.
3. Both use 2. 

I'll refer to these as 1-1, 1-2, and 2-2 for short. Note that of these combinations, 2-2 is the only stable algorithm.

1-1 can be drawn out indefinitely if latency is high, or if the RTT is fairly consistent in both directions. This can be mitigated using randomized exponential backoff, but still relies on timing assumptions to break out of the connection loop.
1-2 is not stable for some pubkey combinations, in particular when A's pubkey is larger than B's. B will continually favor their outbound connection, while A will continue to deny incoming connections from B. (This is why eclair-lnd interop has probably been working for some nodes, but not all)
2-2 _is stable_, as A and B will always fail the same connection. The result is that the protocol always terminates after 1 iteration, and only the losing party needs to teardown/replace their initial connection once.

Ideal Solution
==========
IMO, the ideal solution is that we standardize usage of pukeys as tie-breakers. Of the approaches, is it is the only strategy that terminates determinstically when used with itself.

Unfortunately, there is already a mix of deployed strategies, and in the interrim, the 1-2 combination is also unstable. Thus the remainder of this section is dedicated to formalizing a new strategy that terminates deterministically with 1, 2, and itself. This will promote interoperability in the short term, and can be phased out in favor of 2 over time.

In the meantime, we need something that gracefully handles the 1-2 case. A straightforward fix is to have B deterministically alternate between being honest and deviating from which connection she _should_ drop. B would start in the "honest" state, and flip-flop between the result of the pubkey comparisons. This would reliably terminate after 2 rounds with nodes using 1.

However, nodes implementing this policy can now enter an infinite loop with each other. Observe that this state is volatile, so if one node crashes, the state of each peer may desynchronize. In this case, one of the nodes will always be honest, while the other deviates, resulting in a connection loop.

Proposed Solution
=============
The solution proposed in this PR is a slight deviation on the above strawman. Instead of only using a bit to represent state, we use the integers mod 3. Again, in state 0 a node honors the pubkey comparison. In states 1 and 2, a node deviates and does the opposite.

This change does not require knowing the initial starting state of the nodes. Observe that if the two nodes start in the same state, they will terminate after one round. If they start in different states, the protocol terminates in at most 2 rounds, as it forces a sequence where only one of the peers alters there behavior from the prior round.

Considering the interactions with existing strageies, the worst case for both is 3 rounds:
- Straegy 1. This happens when A's pubkey is smaller and starts in state 1. After returning to state 0, the connection succeeds.
- Straegy 2. This happens when B's pubkey is smaller and  starts in state 1. After returning to state 0, both are running the honest protocol and succeed.

The diff to add this on top of our existing implementation of 2 is quite small, only 16 LOC. I think it'd be great to have strategy 2 formalized moreover as a BOLT, but hoping this will serve as an initial stop gap :)

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -1587,14 +1589,28 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
 // utilize the ordering of the local and remote public key. If we didn't use
 // such a tie breaker, then we risk _both_ connections erroneously being
 // dropped.
-func shouldDropLocalConnection(local, remote *btcec.PublicKey) bool {
+func (s *server) shouldDropLocalConnection(local, remote *btcec.PublicKey) bool {
```
Comment (Roasbeef) [inline server.go]: Does this no longer need these two symptom targeting commits? 

https://github.com/lightningnetwork/lnd/pull/1349/commits/18de5587e679226a641226e94c08df9263ea150a
https://github.com/lightningnetwork/lnd/pull/1349/commits/418267eea9df206e58b0da607f0fef196dc40237

**Comment by Roasbeef** [comment]
> Reject incoming if already have outgoing.

The current code does: "reject incoming if already have incoming". This PR doesn't change that.

File: server.go [hunk 2]
```diff
@@ -1587,14 +1589,28 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
...
 	localPubBytes := local.SerializeCompressed()
 	remotePubPbytes := remote.SerializeCompressed()
 
 	// The connection that comes from the node with a "smaller" pubkey
 	// should be kept. Therefore, if our pubkey is "greater" than theirs, we
 	// should drop our established connection.
-	return bytes.Compare(localPubBytes, remotePubPbytes) > 0
+	supposedToDrop := bytes.Compare(localPubBytes, remotePubPbytes) > 0
```
Comment (Roasbeef) [inline server.go]: `godoc` comment for the method hasn't been updated.

File: server.go (hunk 1 above)
Comment (cfromknecht) [reply server.go]: Nah, still need those. This algo assumes that we are correctly closing only one connection

File: server.go [hunk 3]
```diff
@@ -91,6 +91,7 @@ type server struct {
 	outboundPeers map[string]*peer
 
 	peerConnectedListeners map[string][]chan<- struct{}
+	peerSequence           map[string]byte
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply server.go]: Good question, I think there are cases where we can safely delete and not alter correctness. Will do some research. 

Fwiw we can always delete upon returning to 0

File: server.go (hunk 2 above)
Comment (cfromknecht) [reply server.go]: 👍

**Comment by cfromknecht** [comment]
@roasbeef correct, we should always reject dulicate incoming or duplicate outgoing. For those cases, I’m referring to when a peer has both an inbound and outbound conn. 

That is the startegy (I believe) is employed by eclair, from what I can gather from the Akka docs.

**Comment by cfromknecht** [comment]
Should note that this PR requires #1349

File: server.go [hunk 4]
```diff
@@ -1587,14 +1589,28 @@ func (s *server) peerConnected(conn net.Conn, connReq *connmgr.ConnReq,
...
+	// found, we start in state 0. We take the states mod 3 so that we
+	// the sequence eventually terminates, without making assumptions about
+	// the starting states of the peers.
+	state := s.peerSequence[pubStr]
+	s.peerSequence[pubStr] = (state + 1) % 3
+
+	// Stay honest to the protocol only if we are in state 0.
+	stayHonest := state == 0
+
+	return supposedToDrop == stayHonest
```
Comment (halseth) [inline server.go]: This is very hard to understand without having read the description of the protocol beforehand. Maybe better to write code/comments as  "we choose `A<B` in state 0, and `B>A` in state 1 and 2"?

Comment (cfromknecht) [reply server.go]: If you squint, this is actually computing the xor of the two values. Though it could be documented/explained better :)

**Comment by guggero** [comment]
@ellemouton is this still relevant after the recent changes to the peer connection logic?

**Comment by ellemouton** [comment]
@guggero, the recent changes to peer connection logic handled the tight loop that could occur if we try multiple addresses advertised by one peer in quick succession but does not cover the case where two nodes try connect to each other at the same time which is what i think this PR handles

**Comment by guggero** [comment]
@ellemouton ah, I didn't check the linked issues. You're right, this seems to have been fixed already. Thanks!

Fixed in https://github.com/lightningnetwork/lnd/pull/1349.

**Pull request closed without merging**